src/main/oauth2/utils.ts

```diff
@@ -0,0 +1,148 @@
+    const { authorizationEndpoint, tokenEndpoint } = await fetchAuthorizationServerMetadata();
+    const [code_verifier, code_challenge] = generatePKCE();
+    const redirectHandler = new RedirectHandler();
+    try {
+        const redirect_uri = await redirectHandler.start();
+        const url = createUrlWithQuerystring(authorizationEndpoint, {
+            client_id: OAUTH_CLIENT_ID,
+            response_type: "code",
+            redirect_uri,
+            code_challenge,
+            code_challenge_method: "S256",
+        });
+        openInBrowser(url);
+        const callbackUrl = await redirectHandler.waitForCallback();
+        log.debug(`Received callback URL: ${callbackUrl}`);
+        const code = callbackUrl.searchParams.get("code");
+        log.debug(`Received OAuth2 code: ${code}`);
+        const tokenUrl = createUrlWithQuerystring(tokenEndpoint, {
+            grant_type: "authorization_code",
+            client_id: OAUTH_CLIENT_ID,
+            code,
+            code_verifier,
+            redirect_uri,
+        });
+        const tokenResponse = await fetch(tokenUrl, {
+            method: "POST",
+        });
+        if (tokenResponse.status !== 200) {
+            const error = `Failed to fetch OAuth2 token: ${tokenResponse.status} ${tokenResponse.statusText}`;
+            log.error(error);
+            throw new Error(error);
+        }
+        const { access_token, refresh_token, expires_in } = await tokenResponse.json();
+        if (!access_token || !refresh_token) {
+            const error = "Invalid OAuth2 token response";
+            log.error(error);
+            throw new Error(error);
+        }
+        return {
+            token: access_token,
+            refreshToken: refresh_token,
+            expiresIn: expires_in,
+        };
+    } catch (error) {
+        log.error("Error during login");
+        throw error;
+    } finally {
+        redirectHandler.close();
+    }
+}
+
+export async function renewAccessToken(refreshToken: string): Promise<TokenResponse> {
+    log.debug("Renewing access token");
+    const { tokenEndpoint } = await fetchAuthorizationServerMetadata();
+    if (!refreshToken) {
+        const error = "No refresh token available";
+        log.error(error);
+        stopTokenRenewer();
+        throw new Error(error);
+    }
+    const tokenUrl = createUrlWithQuerystring(tokenEndpoint, {
+        grant_type: "refresh_token",
+        client_id: OAUTH_CLIENT_ID,
+        refresh_token: refreshToken,
+    });
+    const tokenResponse = await fetch(tokenUrl, {
+        method: "POST",
+    });
+    if (tokenResponse.status !== 200) {
+        const error = `Failed to renew token: ${tokenResponse.status} ${tokenResponse.statusText}`;
+        log.error(error);
+        throw new Error(error);
+    }
+    const { access_token, refresh_token, expires_in } = await tokenResponse.json();
+    if (!access_token || !refresh_token) {
```

Comment:
Same as above, refresh token is optional:
> The authorization server MAY issue a new refresh token, in which case
> the client MUST discard the old refresh token and replace it with the
> new refresh token.

at least add a TODO or just a comment that this client requires it.